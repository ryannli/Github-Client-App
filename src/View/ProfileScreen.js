import React, {
  Component
} from 'react';
import {
  AppRegistry,
  StyleSheet,
  ScrollView,
  Text,
  View,
  Dimensions,
  Navigator,
  Linking,
  RefreshControl,
  AsyncStorage
} from 'react-native';
import PTRView from 'react-native-pull-to-refresh';
import PropTypes from 'prop-types';
import {
  Header,
  Title,
  Button,
  Left,
  Right,
  Body,
  List,
  ListItem,
  Container,
  Content,
  Thumbnail
} from 'native-base';
import Icon from 'react-native-vector-icons/SimpleLineIcons'
import {
  StackNavigator,
  SafeAreaView
} from "react-navigation";
import moment from 'moment';
import {
  followUser,
  unfollowUser
} from '../Utils/gitFollow';
import api from '../Utils/basicApi';

// Define style sheet
var styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
  },
  rowContainer: {
    padding: 12
  },
  thumbnail: {
    justifyContent: 'center',
    alignItems: 'center',
    flexGrow: 1,
    height: null,
    width: null,
  },
  headline: {
    textAlign: 'center',
    fontWeight: 'bold',
    fontSize: 18,
    marginTop: 3,
    width: 200,
    fontFamily: 'Palatino',
  },
  username: {
    textAlign: 'center',
    fontSize: 16,
    marginTop: 3,
    width: 200,
    color: '#9E9E9E',
    fontFamily: 'Palatino',
  },
  safeArea: {
    flex: 1,
    // backgroundColor: '#03A9F4',
  },
  follow: {
    textAlign: 'center',
    paddingLeft: 12,
  }
});

export default class Profile extends Component {
  /** Render function. */
  render() {
    var userInfo = this.props.data;

    // The information to retrieve
    var entryArr = ['email', 'bio', 'blog', 'followers', 'following', 'public_repos', 'created_at'];
    // Check if data is ready; otherwise, show "loading" on screen
    var list = userInfo && this.props.login ? entryArr.map((item, index) => {
      return (
        <ListItem icon={true} onPress={() => {this.props.handleClick(item, userInfo[item], this.props.user)}} key={item}>
              <Left>
                <Icon name={this.props.icon[item]} size={22} />
              </Left>
              <Body>
                    <Text>
                      {this.props.screenName[item].toUpperCase()}
                    </Text>
              </Body>
              <Right>
                    <Text>{item == 'created_at'? // Reformat the content to data
                      moment(userInfo[item]).format('LL'):userInfo[item]}</Text>
              </Right>
            </ListItem>
      )
    }) : <Text>loading</Text>;

    // Generates the top avatar, name and username 
    var avatar = userInfo ? <Body style={styles.rowContainer}>     
                                   <Thumbnail large source={{uri: userInfo['avatar_url']}} /> 
                                   <Text style={styles.headline}> {userInfo['name']} </Text>
                                   <Text style={styles.username}> {userInfo['login']} </Text>
                             </Body> :
      <View/>
    var ownerFollowings = this.props.ownerFollowings;
    var actionButtion = <View/>;
    if (userInfo && ownerFollowings && this.props.navigation.state.params.ownername.toLowerCase() != userInfo['login']) {
      if (ownerFollowings.includes(userInfo['login'])) {
        // The user is already followed
        actionButtion = <Button iconLeft block danger onPress={() => {this.props.followAction(false, userInfo['login'])}}>
            <Icon name='user-unfollow' size={25} />
            <Text style={styles.follow}>Unfollow</Text>
          </Button>;
      } else {
        // The user is unfollowed
        actionButtion = <Button iconLeft block success onPress={() => {this.props.followAction(true, userInfo['login'])}}>
            <Icon name='user-follow' size={25} />
            <Text style={styles.follow}>Follow</Text>
          </Button>;
      }

    }

    return (
      <SafeAreaView style={styles.safeArea}>
                <ScrollView refreshControl={
          <RefreshControl
            refreshing={this.props.isRefreshing}
            onRefresh={this.props.onRefresh}
          />
        }>
           <Content style={styles.container}>
            {avatar}
            <List>
              {list}  
            </List>
          </Content>
          {actionButtion}
          </ScrollView>
        </SafeAreaView>
    )
  }
};