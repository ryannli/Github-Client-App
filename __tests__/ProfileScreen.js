import 'react-native';
import React from 'react';
import ProfileScreen from '../src/View/ProfileScreen';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

test('renders correctly', () => {
	const navigation = {
		state: {
			params: {
				username: "ryannli"
			}
		}
	};
	const tree = renderer.create(
		<ProfileScreen navigation={navigation}/>
	);
	expect(tree).toMatchSnapshot();
});

test('renders correctly with data', () => {
	const navigation = {
		state: {
			params: {
				username: "ryannli"
			}
		}
	};
	const data = {
		"login": "ryannli",
		"id": 22183752,
		"avatar_url": "https://avatars2.githubusercontent.com/u/22183752?v=4",
		"gravatar_id": "",
		"url": "https://api.github.com/users/ryannli",
		"html_url": "https://github.com/ryannli",
		"followers_url": "https://api.github.com/users/ryannli/followers",
		"following_url": "https://api.github.com/users/ryannli/following{/other_user}",
		"gists_url": "https://api.github.com/users/ryannli/gists{/gist_id}",
		"starred_url": "https://api.github.com/users/ryannli/starred{/owner}{/repo}",
		"subscriptions_url": "https://api.github.com/users/ryannli/subscriptions",
		"organizations_url": "https://api.github.com/users/ryannli/orgs",
		"repos_url": "https://api.github.com/users/ryannli/repos",
		"events_url": "https://api.github.com/users/ryannli/events{/privacy}",
		"received_events_url": "https://api.github.com/users/ryannli/received_events",
		"type": "User",
		"site_admin": false,
		"name": "Ran Li",
		"company": null,
		"blog": "https://ryannli.github.io/",
		"location": null,
		"email": null,
		"hireable": null,
		"bio": "Love food, sleep and code.",
		"public_repos": 1,
		"public_gists": 0,
		"followers": 3,
		"following": 3,
		"created_at": "2016-09-14T00:14:07Z",
		"updated_at": "2018-03-08T03:05:26Z"
	}
	const tree = renderer.create(
		<ProfileScreen navigation={navigation} data={data}/>
	);
	expect(tree).toMatchSnapshot();
});

test('renders correctly with dictionary features', () => {
	const navigation = {
		state: {
			params: {
				username: "ryannli"
			}
		}
	};
	const icon = {
		"email": "paper-plane",
		"bio": "user",
	}
	const screenName = {
		"email": "email",
		"bio": "bio",
	}
	const tree = renderer.create(
		<ProfileScreen navigation={navigation} icon={icon} screenName={screenName}/>
	);
	expect(tree).toMatchSnapshot();
});