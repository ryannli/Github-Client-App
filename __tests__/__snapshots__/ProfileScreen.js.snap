// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={undefined}
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
    }
  }
>
  <RCTScrollView
    refreshControl={
      <RefreshControlMock
        onRefresh={undefined}
        refreshing={undefined}
      />
    }
  >
    <RCTRefreshControl />
    <View>
      <RCTScrollView
        automaticallyAdjustContentInsets={false}
        contentContainerStyle={undefined}
        contentInset={
          Object {
            "bottom": 0,
          }
        }
        enableAutoAutomaticScroll={true}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        getScrollResponder={[Function]}
        handleOnScroll={[Function]}
        keyboardDismissMode="interactive"
        keyboardOpeningTime={250}
        keyboardShouldPersistTaps="handled"
        keyboardSpace={0}
        onScroll={[Function]}
        resetKeyboardSpace={[Function]}
        resetScrollToCoords={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        scrollEventThrottle={1}
        scrollForExtraHeightOnAndroid={[Function]}
        scrollToEnd={[Function]}
        scrollToFocusedInput={[Function]}
        scrollToPosition={[Function]}
        showsVerticalScrollIndicator={true}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            },
            Object {
              "flex": 1,
              "flexDirection": "column",
            },
          ]
        }
        viewIsInsideTabBar={false}
        virtual={undefined}
      >
        <View>
          <View />
          <View
            closeOnRowBeginSwipe={false}
            closeOnRowPress={true}
            closeOnScroll={true}
            directionalDistanceChangeThreshold={2}
            disableLeftSwipe={false}
            disableRightSwipe={false}
            leftOpenValue={0}
            previewFirstRow={false}
            recalculateHiddenLayout={false}
            rightOpenValue={0}
            style={Object {}}
            swipeToOpenPercent={50}
            virtual={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              loading
            </Text>
          </View>
        </View>
      </RCTScrollView>
      <View />
    </View>
  </RCTScrollView>
</View>
`;

exports[`renders correctly with data 1`] = `
<View
  collapsable={undefined}
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
    }
  }
>
  <RCTScrollView
    refreshControl={
      <RefreshControlMock
        onRefresh={undefined}
        refreshing={undefined}
      />
    }
  >
    <RCTRefreshControl />
    <View>
      <RCTScrollView
        automaticallyAdjustContentInsets={false}
        contentContainerStyle={undefined}
        contentInset={
          Object {
            "bottom": 0,
          }
        }
        enableAutoAutomaticScroll={true}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        getScrollResponder={[Function]}
        handleOnScroll={[Function]}
        keyboardDismissMode="interactive"
        keyboardOpeningTime={250}
        keyboardShouldPersistTaps="handled"
        keyboardSpace={0}
        onScroll={[Function]}
        resetKeyboardSpace={[Function]}
        resetScrollToCoords={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        scrollEventThrottle={1}
        scrollForExtraHeightOnAndroid={[Function]}
        scrollToEnd={[Function]}
        scrollToFocusedInput={[Function]}
        scrollToPosition={[Function]}
        showsVerticalScrollIndicator={true}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            },
            Object {
              "flex": 1,
              "flexDirection": "column",
            },
          ]
        }
        viewIsInsideTabBar={false}
        virtual={undefined}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "flex": 1,
                },
                Object {
                  "padding": 12,
                },
              ]
            }
            virtual={undefined}
          >
            <Image
              large={true}
              source={
                Object {
                  "uri": "https://avatars2.githubusercontent.com/u/22183752?v=4",
                }
              }
              style={
                Object {
                  "borderRadius": 40,
                  "height": 80,
                  "width": 80,
                }
              }
              virtual={undefined}
            />
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "fontFamily": "Palatino",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "marginTop": 3,
                  "textAlign": "center",
                  "width": 200,
                }
              }
            >
               
              Ran Li
               
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#9E9E9E",
                  "fontFamily": "Palatino",
                  "fontSize": 16,
                  "marginTop": 3,
                  "textAlign": "center",
                  "width": 200,
                }
              }
            >
               
              ryannli
               
            </Text>
          </View>
          <View
            closeOnRowBeginSwipe={false}
            closeOnRowPress={true}
            closeOnScroll={true}
            directionalDistanceChangeThreshold={2}
            disableLeftSwipe={false}
            disableRightSwipe={false}
            leftOpenValue={0}
            previewFirstRow={false}
            recalculateHiddenLayout={false}
            rightOpenValue={0}
            style={Object {}}
            swipeToOpenPercent={50}
            virtual={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              loading
            </Text>
          </View>
        </View>
      </RCTScrollView>
      <View />
    </View>
  </RCTScrollView>
</View>
`;

exports[`renders correctly with dictionary features 1`] = `
<View
  collapsable={undefined}
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
    }
  }
>
  <RCTScrollView
    refreshControl={
      <RefreshControlMock
        onRefresh={undefined}
        refreshing={undefined}
      />
    }
  >
    <RCTRefreshControl />
    <View>
      <RCTScrollView
        automaticallyAdjustContentInsets={false}
        contentContainerStyle={undefined}
        contentInset={
          Object {
            "bottom": 0,
          }
        }
        enableAutoAutomaticScroll={true}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        getScrollResponder={[Function]}
        handleOnScroll={[Function]}
        keyboardDismissMode="interactive"
        keyboardOpeningTime={250}
        keyboardShouldPersistTaps="handled"
        keyboardSpace={0}
        onScroll={[Function]}
        resetKeyboardSpace={[Function]}
        resetScrollToCoords={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        scrollEventThrottle={1}
        scrollForExtraHeightOnAndroid={[Function]}
        scrollToEnd={[Function]}
        scrollToFocusedInput={[Function]}
        scrollToPosition={[Function]}
        showsVerticalScrollIndicator={true}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            },
            Object {
              "flex": 1,
              "flexDirection": "column",
            },
          ]
        }
        viewIsInsideTabBar={false}
        virtual={undefined}
      >
        <View>
          <View />
          <View
            closeOnRowBeginSwipe={false}
            closeOnRowPress={true}
            closeOnScroll={true}
            directionalDistanceChangeThreshold={2}
            disableLeftSwipe={false}
            disableRightSwipe={false}
            leftOpenValue={0}
            previewFirstRow={false}
            recalculateHiddenLayout={false}
            rightOpenValue={0}
            style={Object {}}
            swipeToOpenPercent={50}
            virtual={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              loading
            </Text>
          </View>
        </View>
      </RCTScrollView>
      <View />
    </View>
  </RCTScrollView>
</View>
`;
